presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.operator.aggregation.InternalAggregationFunction;
+import com.facebook.presto.operator.scalar.ScalarFunctionImplementation;
+import com.facebook.presto.operator.window.WindowFunctionSupplier;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.BlockEncodingSerde;
+import com.facebook.presto.spi.function.OperatorType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.sql.SqlPathElement;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class FunctionManager
+{
+    private final FunctionNamespace globalFunctionNamespace;
+    private volatile Map<SqlPathElement, FunctionNamespace> dynamicFunctionNamespaces;
+    private final TypeManager typeManager;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final FeaturesConfig featuresConfig;
+
+    public FunctionManager(TypeManager typeManager, BlockEncodingSerde blockEncodingSerde, FeaturesConfig featuresConfig)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        this.globalFunctionNamespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+        this.dynamicFunctionNamespaces = ImmutableMap.of();
+    }
+
+    public synchronized void addFunctions(String catalog, String schema, List<? extends SqlFunction> functions)
+    {
+        if (!dynamicFunctionNamespaces.containsKey(catalog)) {
```

Comment:
Now that `dynamicFunctionNamespaces` is a `Map<SqlPathElement, FunctionNamespace>`, you should fix the various `containsKey` and `get` calls here and below as they pass `String` to those methods instead of `SqlPathElement`.